File: pbark/pbark/parser/collection/pile_action_parser.py
from __future__ import annotations

from pbark.config.config_loader import ConfigLoader
from pbark.errors import BarkError
from pbark.lexer import TokenType
from pbark.parser import ast_node as AstNode
from pbark.parser import keywords as Keywords
from pbark.parser.parse_expression import NumberLiteral, StringLiteral


class PileActionParser:
    def __init__(self, parser) -> None:
        self._parser = parser

    def parse_grab_top_line(self):
        pile = self._find_pile_after_from()
        if pile is None or not self._parser.line_has_word(Keywords.PILE_GRAB_FROM_WORDS):
            return None
        line = self._parser.peek().line
        if not self._line_has_top_before_from():
            raise BarkError(
                line,
                "Grab what from the pile? The dogs need to know which item they can have!",
            )
        subject = self._find_dog_on_line()
        if subject is None:
            raise BarkError(
                line, "Who is grabbing from the pile? Name the dog or use she/he first"
            )
        ConfigLoader.require_pile(pile, line)
        self._parser.consume_until_line_end()
        return AstNode.PilePop(pile, line)

    def parse_push_into_line(self):
        into_at = self._find_in_or_into_offset()
        if into_at < 0:
            return None
        pile = ConfigLoader.resolve_collection_from_tokens(
            self._parser.all_tokens(),
            self._parser.cursor_index() + into_at + 1,
            self._parser.line_end_index(),
        )
        if pile is None or not ConfigLoader.is_pile(pile):
            return None
        if not self._parser.line_has_word(Keywords.PILE_PUSH_INTO_WORDS):
            return None
        line = self._parser.peek().line
        item = self._find_item_before_in(into_at)
        if item is None:
            raise BarkError(line, "Toss what into the pile? The dogs are curious and want to know!")
        subject = self._find_dog_on_line()
        if subject is None:
            raise BarkError(
                line,
                "Who is putting something in the pile? Name the dog or use she/he first",
            )
        ConfigLoader.require_pile(pile, line)
        self._parser.consume_until_line_end()
        return AstNode.PilePush(pile, item, line)

    def _find_pile_after_from(self) -> str | None:
        line_start = self._parser.cursor_index()
        line_end = self._parser.line_end_index()
        for from_at in range(line_start, line_end):
            token = self._parser.all_tokens()[from_at]
            if not token.is_(TokenType.IDENTIFIER):
                continue
            if self._parser.normalise(token.value) not in Keywords.FROM_WORDS:
                continue
            pile = ConfigLoader.resolve_collection_from_tokens(
                self._parser.all_tokens(), from_at + 1, line_end
            )
            if pile is not None and ConfigLoader.is_pile(pile):
                return pile
        return None

    def _line_has_top_before_from(self) -> bool:
        line_start = self._parser.cursor_index()
        line_end = self._parser.line_end_index()
        for from_at in range(line_start, line_end):
            token = self._parser.all_tokens()[from_at]
            if not token.is_(TokenType.IDENTIFIER) or self._parser.normalise(token.value) not in Keywords.FROM_WORDS:
                continue
            for offset in range(from_at - 1, line_start - 1, -1):
                before = self._parser.all_tokens()[offset]
                if not before.is_(TokenType.IDENTIFIER):
                    continue
                if self._parser.normalise(before.value) in Keywords.TOP_WORDS:
                    return True
            return False
        return False

    def _find_in_or_into_offset(self) -> int:
        offset = 0
        while not self._parser.is_at_end_or_line_end_offset(offset):
            token = self._parser.peek_at(offset)
            if not token.is_(TokenType.IDENTIFIER):
                offset += 1
                continue
            if self._parser.normalise(token.value) in Keywords.IN_WORDS:
                return offset
            offset += 1
        return -1

    def _find_item_before_in(self, into_at: int):
        for offset in range(into_at - 1, -1, -1):
            token = self._parser.peek_at(offset)
            if token.is_(TokenType.STRING):
                return StringLiteral(token.value)
            if token.is_(TokenType.NUMBER):
                return NumberLiteral(float(self._parser.normalise(token.value)))
        return None

    def _find_dog_on_line(self) -> str | None:
        offset = 0
        while not self._parser.is_at_end_or_line_end_offset(offset):
            token = self._parser.peek_at(offset)
            if not token.is_(TokenType.IDENTIFIER):
                offset += 1
                continue
            word = self._parser.normalise(token.value)
            if Keywords.is_dog_subject_reference(word, token.value):
                return word
            offset += 1
        return None

File: src/main/java/dev/klomptech/jbark/parser/collection/PileActionParser.java
package dev.klomptech.jbark.parser.collection;

import dev.klomptech.jbark.config.ConfigLoader;
import dev.klomptech.jbark.errors.BarkError;
import dev.klomptech.jbark.lexer.Token;
import dev.klomptech.jbark.lexer.TokenType;
import dev.klomptech.jbark.parser.AstNode;
import dev.klomptech.jbark.parser.Keywords;
import dev.klomptech.jbark.parser.ParseExpression;
import dev.klomptech.jbark.parser.Parser;

public class PileActionParser {

  private final Parser parser;

  public PileActionParser(final Parser parser) {
    this.parser = parser;
  }

  // Example: "she takes the top item from the laundry basket" (top item comes off the stack)
  public AstNode parseGrabTopLine() throws BarkError {
    String pile = findPileAfterFrom();
    if (pile == null || !parser.lineHasWord(Keywords.PILE_GRAB_FROM_WORDS)) {
      return null;
    }
    int line = parser.peek().line();
    if (!lineHasTopBeforeFrom()) {
      throw new BarkError(
          line, "Grab what from the pile? The dogs need to know which item they can have!");
    }
    String subject = findDogOnLine();
    if (subject == null) {
      throw new BarkError(line, "Who is grabbing from the pile? Name the dog or use she/he first");
    }
    ConfigLoader.requirePile(pile, line);
    parser.consumeUntilLineEnd();
    return new AstNode.PilePop(pile, line);
  }

  // Example: "she puts "slipper" into the laundry basket" (one item goes on top)
  public AstNode parsePushIntoLine() throws BarkError {
    int intoAt = findInOrIntoOffset();
    if (intoAt < 0) {
      return null;
    }
    String pile =
        ConfigLoader.resolveCollectionFromTokens(
            parser.allTokens(), parser.cursorIndex() + intoAt + 1, parser.lineEndIndex());
    if (pile == null || !ConfigLoader.isPile(pile)) {
      return null;
    }
    if (!parser.lineHasWord(Keywords.PILE_PUSH_INTO_WORDS)) {
      return null;
    }
    int line = parser.peek().line();
    ParseExpression item = findItemBeforeIn(intoAt);
    if (item == null) {
      throw new BarkError(line, "Toss what into the pile? The dogs are curious and want to know!");
    }
    String subject = findDogOnLine();
    if (subject == null) {
      throw new BarkError(
          line, "Who is putting something in the pile? Name the dog or use she/he first");
    }
    ConfigLoader.requirePile(pile, line);
    parser.consumeUntilLineEnd();
    return new AstNode.PilePush(pile, item, line);
  }

  private String findPileAfterFrom() {
    int lineStart = parser.cursorIndex();
    int lineEnd = parser.lineEndIndex();
    for (int fromAt = lineStart; fromAt < lineEnd; fromAt++) {
      Token token = parser.allTokens().get(fromAt);
      if (!token.is(TokenType.IDENTIFIER)) {
        continue;
      }
      if (!Keywords.FROM_WORDS.contains(parser.normalise(token.value()))) {
        continue;
      }
      String pile =
          ConfigLoader.resolveCollectionFromTokens(parser.allTokens(), fromAt + 1, lineEnd);
      if (pile != null && ConfigLoader.isPile(pile)) {
        return pile;
      }
    }
    return null;
  }

  private boolean lineHasTopBeforeFrom() {
    int lineStart = parser.cursorIndex();
    int lineEnd = parser.lineEndIndex();
    for (int fromAt = lineStart; fromAt < lineEnd; fromAt++) {
      Token token = parser.allTokens().get(fromAt);
      if (!token.is(TokenType.IDENTIFIER)
          || !Keywords.FROM_WORDS.contains(parser.normalise(token.value()))) {
        continue;
      }
      for (int offset = fromAt - 1; offset >= lineStart; offset--) {
        Token before = parser.allTokens().get(offset);
        if (!before.is(TokenType.IDENTIFIER)) {
          continue;
        }
        String word = parser.normalise(before.value());
        if (Keywords.TOP_WORDS.contains(word)) {
          return true;
        }
      }
      return false;
    }
    return false;
  }

  private int findInOrIntoOffset() {
    for (int offset = 0; !parser.isAtEndOrLineEndOffset(offset); offset++) {
      Token token = parser.peekAt(offset);
      if (!token.is(TokenType.IDENTIFIER)) {
        continue;
      }
      String word = parser.normalise(token.value());
      if (Keywords.IN_WORDS.contains(word)) {
        return offset;
      }
    }
    return -1;
  }

  private ParseExpression findItemBeforeIn(final int intoAt) {
    for (int offset = intoAt - 1; offset >= 0; offset--) {
      Token token = parser.peekAt(offset);
      if (token.is(TokenType.STRING)) {
        return new ParseExpression.StringLiteral(token.value());
      }
      if (token.is(TokenType.NUMBER)) {
        return new ParseExpression.NumberLiteral(
            Double.parseDouble(parser.normalise(token.value())));
      }
    }
    return null;
  }

  private String findDogOnLine() {
    for (int offset = 0; !parser.isAtEndOrLineEndOffset(offset); offset++) {
      Token token = parser.peekAt(offset);
      if (!token.is(TokenType.IDENTIFIER)) {
        continue;
      }
      String word = parser.normalise(token.value());
      if (Keywords.isDogSubjectReference(word, token.value())) {
        return word;
      }
    }
    return null;
  }
}
